### PR TITLE
Support repeat at any dimension

### PR DESCRIPTION
**Pull request by johnnychen94**
This solves issue #29614 as an enhancement

now `repeat(ones(2,2), 2,2,2)` won't raise `BoundsError` but instead return a `Array{T,3}` as intuition expects. 

Benchmark:
```julia
julia> a = ones(100,100,100);
# before this patch
julia> @btime repeat(a, 2,2,2);
  81.971 ms (40 allocations: 114.44 MiB)
# after this patch
julia> @btime repeat(a, 2,2,2);
  82.136 ms (40 allocations: 114.44 MiB)
```

### CODE REVIEW

File: base/abstractarraymath.jl [hunk 1]
```diff
@@ -377,7 +377,8 @@ _reperr(s, n, N) = throw(ArgumentError("number of " * s * " repetitions " *
 
     # fill the first inner block
     if all(x -> x == 1, inner)
-        R[axes(A)...] = A
+        idxs = (axes(A)..., ntuple(n->OneTo(1), length(shape) - length(inner))...) # keep dimension consistent
```
Comment (mbauman) [inline base/abstractarraymath.jl]: Rather than `OneTo(1)`, it might be slightly more friendly to bizarro array subtypes to just use `:`s here.

**Comment by mbauman** [comment]
Welcome and thanks for the contribution!

File: base/abstractarraymath.jl (hunk 1 above)
Comment (johnnychen94) [reply base/abstractarraymath.jl]: I was trying to figure out a simple way of getindex. Got it!

File: base/abstractarraymath.jl [hunk 2]
```diff
@@ -377,7 +377,7 @@ _reperr(s, n, N) = throw(ArgumentError("number of " * s * " repetitions " *
 
     # fill the first inner block
     if all(x -> x == 1, inner)
-        R[axes(A)...] = A
+        R[axes(A)...,:] = A
```
Comment (mbauman) [inline base/abstractarraymath.jl]: Ah, sorry, I wasn't clear!  You'll still need the appropriately sized tuple like you had before, but just a tuple filled with `:`s instead of `OneTo(1)`s.

Comment (johnnychen94) [reply base/abstractarraymath.jl]: Well...
Looks like both
```julia
R[axes(A)...,:] = A
```
and
```julia
idxs = (axes(A)..., ntuple(n->:, length(shape) - length(inner))...)
R[idxs...] = A
```
throws 
`DimensionMismatch("tried to assign 2×2 array to 2×2×2 destination")`
for `repeat(ones(2,2), 2, 2, 2)`

I think I'll just rollback to the original one.

Comment (johnnychen94) [reply base/abstractarraymath.jl]: The problem here is that `R` is always a larger array than `A`.
For outer repeat such as `repeat(ones(2,2), 2, 2, 2)`, this line of code simply fills the initial pixels with known values, leaving other pixels still undef. So we can't use `:` in this case.

**Comment by johnnychen94** [comment]
Thanks for the reviewing @mbauman and sorry for the back-and-forth commits on the same code.

I am not very familiar with the index mechanisms of Julia Arrays; I thought `1` is equivalent to `OneTo(1)` in this case, but the test says no.

I was even thinking of using 
```julia
R[axes(A)...,ones(Int, length(shape) - length(inner))...] = A
```
to replace
```julia
idxs = (axes(A)..., ntuple(n->OneTo(1), length(shape) - length(inner))...)
R[idxs...] = A
```
But not very sure if this works. (Actually I don't know how to fully test it in my own machine)

**Comment by mbauman** [comment]
The best way to test changes like this locally is to use Revise.jl, but sometimes it's quickest/easiest to [just `@eval` the code in the REPL](https://gist.github.com/mbauman/e5d35b780aed46d6ab69c79a31484446).  Just type `@eval Base` (where `Base` is the module in which the original code was defined) and then paste the new definition.

Sorry for the misdirection on the colons — of course we just want to be assigning into the _first_ block.  Using `1` instead of `OneTo(1)` does indeed behave the same in `setindex!` and does indeed work… you just got bit by a spurious network interruption in an unrelated test on CI.  I do find `1` to be a bit simpler and clearer in intent, but it's not a big deal.  More importantly, though, I think you want to just use `ndims(R) - ndims(A)` instead of the lengths of inner and shape.  Could you add this test case?

```julia
repeat(ones(2,2), inner=(1, 1, 1), outer=(2, 2, 2))
```

With respect to your question about `ones`: the challenge is doing these index manipulations in a way that is type-stable — that is, in a way where Julia can infer the length at compile time.  Splatting an array like `ones(…)` into a tuple is definitely not what you want in that case as it allocates an array and Julia doesn't store array lengths in the type system.  Using `ntuple` is better — especially if we can use an inferable constant for the second argument… but that subtraction isn't inferable.  It looks like this whole function could use a bit of hand optimizing… so making this better here can be future work.  Also good future work could be making this work in cases where there is an inner repeat.

**Comment by johnnychen94** [comment]
Thanks for this explanation, this helps me a lot.

With regard to optimization,  I did some relative benchmark tests as a reference (kind of interesting to me). This benchmark says, at least, we can optimize `repeat` for some special cases.

```julia
# Baseline
# without this patch
julia> @btime repeat(ones(100,100), 20, 20);
  8.828 ms (4 allocations: 30.59 MiB)

# without this patch (slightly faster than Baseline)
julia> @btime reshape(repeat(ones(100,100),20,20),2000,2000,1,1,1);
  8.590 ms (6 allocations: 30.59 MiB)

# without this patch
julia> @btime repeat(reshape(ones(100,100),100,100,1,1,1),20,20,1,1,1);
66.962 ms (3625 allocations: 61.17 MiB)

# with this patch
julia> @btime repeat(ones(100,100), 20, 20,1,1,1);
70.049 ms (3625 allocations: 61.17 MiB)
```

I'll tell you when I have a solution to support this feature without loss much performance. (I have some other jobs to do, so no sooner than Nov)
But still, I think it's okay to merge this now and recreate a new PR when we find a way to optimize this.

**Comment by johnnychen94** [comment]
ping @mbauman I think this is good to merge. Can you have a quick look?

**Comment by mbauman** [comment]
Yes, you're exactly right and this did indeed work previously. It had the somewhat strange behavior that the trailing index was always allowed to "linearly span" (working through memory order for plain old `Array`) over any omitted dimensions.  This ran us into trouble in a number of situations, including consistency with offset arrays, the meaning of the `end` keyword, and expectations of bounds errors. For more details see: #14470, #20079, and #21750.  We now have the following rules:

* Indexing with one index is linear indexing
* Indexing with less than `ndims` indices is only supported if the dimensions omitted are all "singleton" (e.g., length 1) 
* Indexing with more than `ndims` indices is only supported if the extra dimensions provided are all selecting the first index (e.g., 1)

In other words, you've found and cleaned up some of my own dirty laundry — thank you!

**Comment by johnnychen94** [comment]
Cool, I'll keep this in mind.
